src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/CompletionTriggerAndCommitCharacters.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Razor.Completion;
+
+internal class CompletionTriggerAndCommitCharacters
+{
+    /// <summary>
+    ///  Trigger character that can trigger both Razor and Delegation completion
+    /// </summary>
+    private const char TransitionCharacter = '@';
+
+    private static readonly char[] s_vsHtmlTriggerCharacters = [':', '#', '.', '!', '*', ',', '(', '[', '-', '<', '&', '\\', '/', '\'', '"', '=', ':', ' ', '`'];
+    private static readonly char[] s_vsCodeHtmlTriggerCharacters = ['#', '.', '!', ',', '-', '<'];
+    private static readonly char[] s_razorTriggerCharacters = ['<', ':', ' '];
+    private static readonly char[] s_csharpTriggerCharacters = [' ', '(', '=', '#', '.', '<', '[', '{', '"', '/', ':', '~'];
+    private static readonly char[] s_commitCharacters = [' ', '>', ';', '='];
+
+    private readonly HashSet<char> _csharpTriggerCharacters;
+    private readonly HashSet<char> _delegationTriggerCharacters;
+    private readonly HashSet<char> _htmlTriggerCharacters;
+    private readonly HashSet<char> _razorTriggerCharacters;
+    private readonly ImmutableArray<string> _allTriggerCharacters;
+    private readonly ImmutableArray<string> _allCommitCharacters;
+
+    // Note: Create a new array each time to avoid modifying the original array.
+    public string[] AllTriggerCharacters => [.. _allTriggerCharacters];
```

Comment:
`_allTriggerCharacters` is already immutable. Is there a reason not to expose a new array as mutable here?

Reply:
There is a reason, but it's admittedly not a great one. The only callers immediately assign the value into `CompletionRegistrationOptions.TriggerCharacters`, which requires a mutable array since it's an LSP object. However, it probably makes more sense to just return the immutable array here and convert it to a mutable array at the call site. I'll do that. Thanks for asking about this!

Reply:
This should be better now.